Fix compatibility with UnifiedPush/tink

p1gp1g opened the pull request:
Bring back openssl3 support and fix encryption

p1gp1g committed:
Fix encryption, following RFC8291

ClearlyClaire commented on lib/webpush/encryption.rb:
```diff
@@ -7,12 +7,14 @@ module Encryption
     # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
     def encrypt(message, p256dh, auth)
       assert_arguments(message, p256dh, auth)
+      # Following RFC8291, messages can't be longer than 3993 bytes long
+      # so encrypted message do not exceed 4096 bytes
+      raise ArgumentError, "message is too big" if message.bytesize > 3993
```
If I understand correctly, this is just an early check for the `> 4096` that happens later, right?

p1gp1g replied:
Yes, doing this early check avoid doing other instructions for nothing

ClearlyClaire commented on lib/webpush/encryption.rb:
```diff
@@ -48,13 +55,14 @@ def encrypt(message, p256dh, auth)
     def encrypt_payload(plaintext, content_encryption_key, nonce)
+      # RFC8291 requires the padding delimiter to be 0x02
+      plaintext = plaintext + "\x02"
       cipher = OpenSSL::Cipher.new('aes-128-gcm')
       cipher.encrypt
       cipher.key = content_encryption_key
       cipher.iv = nonce
       text = cipher.update(plaintext)
-      padding = cipher.update("\2\0")
-      e_text = text + padding + cipher.final
+      e_text = text + cipher.final
       e_tag = cipher.auth_tag
```
If I understand correctly, the previous version was correct, but added one byte of unnecessary padding, right?

p1gp1g replied:
Yes, but:
- that may be the reason why decryption failed with other libs (such as [tink](https://github.com/tink-crypto/tink-java-apps/tree/main/webpush) used by [UnifiedPush lib](https://codeberg.org/UnifiedPush/android-connector)
- it is impossible to test [against the RFC example](https://www.rfc-editor.org/rfc/rfc8291#appendix-A) with that padding

ClearlyClaire commented on lib/webpush/encryption.rb:
```diff
@@ -34,10 +36,15 @@ def encrypt(message, p256dh, auth)
 
       ciphertext = encrypt_payload(message, content_encryption_key, nonce)
+      # This should never happen if message length <= 3993
+      raise ArgumentError, "encrypted payload is too big" if ciphertext.bytesize > 4096
 
       serverkey16bn = convert16bit(server_public_key_bn)
-      rs = ciphertext.bytesize
-      raise ArgumentError, "encrypted payload is too big" if rs > 4096
+      # According to RFC8188, the final record can be smaller than the record size
+      # RFC8291 requires encrypted messages to be at most 4096. And the example set the
+      # RS to 4096. RS can be smaller to 4096 but hardcoding is also following the specs.
+      # We set RS=4096 to allow testing with the RFC example.
+      rs = 4096
```
This is one thing I'm a bit confused by, though I'm not really knowledgeable about aes128gcm. Why is `rs` now unrelated to the actual record size?

p1gp1g replied:
RS gives the maximum size of the record, RFC8291 tells to use 4096 (it should be possible to use a lower rs, but it isn't necessary, and breaks the example. It may be possible that this rs was the cause of the decryption failure with tink lib)

ClearlyClaire replied:
RFC8291 does not say to use 4096 as `rs`, it only says the following:
> An application server MUST set the "rs" parameter in the "aes128gcm" content coding header to a size that is greater than the sum of the lengths of the plaintext, the padding delimiter (1 octet), any padding, and the authentication tag (16 octets).

So basically that it should be enough to hold the value of `ciphertext`.

p1gp1g replied:
> An application server MUST set the "rs" parameter in the "aes128gcm" content coding header to __a size that is greater than__ the sum of the lengths of the plaintext, the padding delimiter (1 octet), any padding, and the authentication tag (16 octets).

It means it can always be 4096, but may be lower. That is what they do with the example:

>    The salt, record size of 4096, and application server public key
   produce an 86-octet header of:
>
>   DGv6ra1nlYgDCS1FRnbzlwAAEABBBP4z 9KsN6nGRTbVYI_c7VJSPQTBtkgcy27ml
   mlMoZIIgDll6e3vCYLocInmYWAmS6Tlz AC8wEqKK6PBru3jl7A8

And being able to reproduce the example is IMO very important for a library implementing an RFC

ClearlyClaire reviewed (commented):
Hi! Thank you for your contribution.

First, can you rebase this PR? `main` should have the OpenSSL 3 fixes already.

I'm not very knowledgeable about RFC8291 and aes128gcm, but my understanding is that the current code is already correct, and your changes amount to:
- checking message size before encryption for earlier error handling
- removing an unnecessary byte of padding after the padding delimiter
- choosing a fixed value of 4096 for the record size

p1gp1g commented:
> Hi! Thank you for your contribution.
> 
> First, can you rebase this PR? `main` should have the OpenSSL 3 fixes already.

I've rebased the branch

> I'm not very knowledgeable about RFC8291 and aes128gcm, but my understanding is that the current code is already correct, and your changes amount to:
> 
>     * checking message size before encryption for earlier error handling
> 
>     * removing an unnecessary byte of padding after the padding delimiter
> 
>     * choosing a fixed value of 4096 for the record size

It is currently not possible to decrypt push messages with other libs, the error may come from that extra padding or from a wrong RS.

Removing that extra padding and using the fixed RS fixes decryption with other libs (tested with [tink](https://github.com/tink-crypto/tink-java-apps/tree/main/webpush) used by [UnifiedPush lib](https://codeberg.org/UnifiedPush/android-connector)), and also allow testing against the RFC example (that is what I've implemented in mastodon PR)

ClearlyClaire commented:
Right, so this PR changes various parameters to use the particular ones used in the RFC example to be able to use it as a verifiable example.

As for `tink` failing to process our messages without changes, I was not aware of that and will look into it, as it sounds like an issue with `tink`.

p1gp1g commented:
I haven't check if there were an issue on tink side

ClearlyClaire commented:
Looking at `tink`'s code, I believe the extra padding should be fine:
https://github.com/tink-crypto/tink-java-apps/blob/4295187cb1c8b05d01146c1a6b5fb45d47ed7fd5/webpush/src/main/java/com/google/crypto/tink/apps/webpush/WebPushHybridDecrypt.java#L271-L278

However, if my understanding is correct, the record size may be the issue, as it is set to 4096 unless explicitly defined otherwise:
https://github.com/tink-crypto/tink-java-apps/blob/4295187cb1c8b05d01146c1a6b5fb45d47ed7fd5/webpush/src/main/java/com/google/crypto/tink/apps/webpush/WebPushHybridDecrypt.java#L135
https://github.com/tink-crypto/tink-java-apps/blob/4295187cb1c8b05d01146c1a6b5fb45d47ed7fd5/webpush/src/main/java/com/google/crypto/tink/apps/webpush/WebPushHybridDecrypt.java#L148-L151

And checked against what is read from the record:
https://github.com/tink-crypto/tink-java-apps/blob/4295187cb1c8b05d01146c1a6b5fb45d47ed7fd5/webpush/src/main/java/com/google/crypto/tink/apps/webpush/WebPushHybridDecrypt.java#L230-L235

https://codeberg.org/UnifiedPush/android-connector does not set the `recordSize`, so it would default to 4096 and fail the validation.

So my belief is that our current code is correct and UnifiedPush/tink fails to decode it because of an superfluous check.

p1gp1g commented:
Maybe the decryption was correct and the error come from that RS indeed :+1: 

The extra padding could let message with exact length of 3993 to fail even if it could succeed

p1gp1g commented:
@ClearlyClaire : There was indeed a bug in tink: https://github.com/tink-crypto/tink-java-apps/pull/5
